bluemira/equilibria/coils.py

```diff
@@ -85,27 +92,318 @@ def generate_name(coil: Any, idx: Optional[int] = None):
+        method is used for the field inside the coil.
+
+        Parameters
+        ----------
+        x: Union[float, int, np.array]
+            The x values at which to calculate the Bz response
+        z: Union[float, int, np.array]
+            The z values at which to calculate the Bz response
+
+        Returns
+        -------
+        Bz: Union[float, np.array]
+            The vertical magnetic field response at the x, z coordinates.
+        """
+        return self._mix_control_method(
+            x, z, self._control_Bz_greens, self._control_Bz_analytical
+        )
+
+    def Bp(self, x, z):
+        """
+        Calculate poloidal magnetic field Bp at (x, z)
+        """
+        return np.hypot(self.Bx(x, z), self.Bz(x, z))
+
+    def _mix_control_method(self, x, z, greens_func, semianalytic_func):
+        """
+        Boiler-plate helper function to mixed the Green's function responses
+        with the semi-analytic function responses, as a function of position
+        outside/inside the coil boundary.
+        """
+        x, z = np.ascontiguousarray(np.atleast_2d(x)), np.ascontiguousarray(
+            np.atleast_2d(z)
+        )
+
+        lg_or = np.logical_or(self._quad_dx == 0, self._quad_dz == 0)
+
+        if False in lg_or:
+            # if dx or dz is not 0 and x,z inside coil
+            # TODO improve to remove inside coil calc if already known
+            inside = np.logical_and(self._points_inside_coil(x, z), not lg_or)
+
+            response = np.zeros(x.shape)
+
+            resp = response.reshape(-1)
+            ins = inside.reshape(-1)
+
+            if np.any(~inside):
+                resp[~ins] = greens_func(x[~inside], z[~inside]).reshape(-1)
+            if np.any(inside):
+                resp[ins] = semianalytic_func(x[inside], z[inside]).reshape(-1)
+        else:
+            response = greens_func(x, z)
+
+        return response
+
+    def _points_inside_coil(self, x, z):
+        """
+        Determine which points lie inside or on the coil boundary.
+
+        Parameters
+        ----------
+        x: Union[float, np.array]
+            The x coordinates to check
+        z: Union[float, np.array]
+            The z coordinates to check
+
+        Returns
+        -------
+        inside: np.array(dtype=bool)
+            The Boolean array of point indices inside/outside the coil boundary
+        """
+        x, z = np.ascontiguousarray(x), np.ascontiguousarray(z)
+        # Add an offset, to ensure points very near the edge are counted as
+        # being on the edge of a coil
+        atol = X_TOLERANCE
+        x_min, x_max = (
+            self._quad_x - self._quad_dx - atol,
+            self._quad_x + self._quad_dx + atol,
+        )
+        z_min, z_max = (
+            self._quad_z - self._quad_dz - atol,
+            self._quad_z + self._quad_dz + atol,
+        )
+        return (x >= x_min) & (x <= x_max) & (z >= z_min) & (z <= z_max)
```

Comment:
Is this still just checking whether the point is inside the full coil? Looks like it might be checking on all the quad points

Reply:
It is point inside a given quadrature. For most B field calculations I thought that would be what you'd want because if _quad_dz or _quad_dx = 0 then we can just use the greens function?